scipy/optimize/_minimize.py

```diff
@@ -601,54 +601,99 @@ def minimize(fun, x0, args=(), method=None, jac=None, hess=None,
                       bounds=bounds, constraints=constraints,
                       callback=callback, **options)
 
-    if bounds is not None:
-        bounds = standardize_bounds(bounds, x0, meth)
-
     if constraints is not None:
         constraints = standardize_constraints(constraints, x0, meth)
 
+    remove_vars = False
+    if bounds is not None:
+        if meth.upper() in {"TNC", "SLSQP", "L-BFGS-B"}:
+            # These methods can't take the finite-difference derivatives they
+            # need when a variable is fixed by the bounds. To avoid this issue,
+            # remove fixed variables from the problem.
+
+            # convert to new-style bounds so we only have to consider one case
+            bounds = standardize_bounds(bounds, x0, 'new')
+
+            # determine whether any variables are fixed
+            i_fixed = (bounds.lb == bounds.ub)
+
+            # determine whether finite differences are needed for any grad/jac
+            fd_needed = not(callable(jac))
+            for con in constraints:
```

Comment:
I think `constraints` can still be `None` at this point so this for loop would fail with:
```
TypeError: 'NoneType' object is not iterable
```

Reply:
I'll investigate, and no harm in making it work with `None` if it works in master. (But it's not a documented legal value. The default is an empty tuple.)

Reply:
You're correct, the default for `constraints` is an empty tuple. I guess I was being lead by existing code that checks that bounds and constraints are not None:

```
    if bounds is not None:
        bounds = standardize_bounds(bounds, x0, meth)

    if constraints is not None:
        constraints = standardize_constraints(constraints, x0, meth)
```

Reply:
Done.